src/lib/storage/matches_all_pos_list.hpp

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include "./chunk.hpp"
+#include "abstract_pos_list.hpp"
+
+namespace opossum {
+
+class MatchesAllPosList : public AbstractPosList {
+ public:
+  explicit MatchesAllPosList(std::shared_ptr<const Chunk> common_chunk, const ChunkID common_chunk_id)
+      : _common_chunk(common_chunk),
+        _common_chunk_id(common_chunk_id),
+        _common_chunk_size_on_creation(common_chunk->size()) {}
```

Comment:
We would vote for adding a check inside the constructor that ```common_chunk``` is final.
This should catch most of the errors, but we think some of them have to be handled outside the MatchesAllPoslist.

Consider following (constructed) scenario:
---------------
```
                    A                                                B

+------------------------------------------------+-----------------------------------+
                                                 |
 Scans all rows and sees that everything matches |
                                                 |
                                                 |  Adds rows and finalizes chunk
                                                 |
  Create MatchesAllPosList with finalized chunk  |
(containing too many Rows)                       |
                                                 +

```
--------------
This can only be eliminated by A checking before the scan (we think)

Reply:
What if the constructor simply took the chunk size at the time of scanning / validating / ...? If instead of

```
  explicit MatchesAllPosList(std::shared_ptr<const Chunk> common_chunk, const ChunkID common_chunk_id)
      : _common_chunk(common_chunk),
        _common_chunk_id(common_chunk_id),
        _common_chunk_size_on_creation(common_chunk->size()) {}
```

we had

```
  explicit MatchesAllPosList(const ChunkID common_chunk_id, const ChunkOffset common_chunk_size)
      : _common_chunk_id(common_chunk_id),
        _common_chunk_size(chunk_size) {}
```

this should not be an issue. It should also be easier for users. `_common_chunk` seems unused anyway.

Reply:
I implemented what you suggested, but I'm not sure about the thread safety (I think either master is broken already or our code that stores the size when the ctor is called should be safe)

In the current master, the algorithm for generating the pos list when taking all shortcuts is:
```c++
if (_can_use_chunk_shortcut && _is_entire_chunk_visible(chunk_in, snapshot_commit_id)) {
  // boom
  const auto chunk_size = chunk_in->size();
  for (auto chunk_offset = 0u; chunk_offset < chunk_size; ++chunk_offset) {
    temp_pos_list[chunk_offset] = RowID{chunk_id, chunk_offset};
  }
}
```
and `temp_pos_list` is then moved to the result PosList. However, where I added `// boom`, we are already in a race condition, or is there some kind of transparent locking going on that I'm missing? If, between checking whether the whole chunk matches, and then getting the size, a new value is added that does not match, then we will include it erroneously, right?